pkg/controller/clustercatalog_controller.go

```diff
@@ -0,0 +1,156 @@
+}
+
+func defaultApplyFunc(client dynamic.Interface) applyFunc {
+	return func(ctx context.Context, key types.NamespacedName, fieldManager string, force bool, gvr schema.GroupVersionResource, manifest []byte) error {
+		var resourceInterface dynamic.ResourceInterface = client.Resource(gvr)
+		if key.Namespace != "" {
+			resourceInterface = client.Resource(gvr).Namespace(key.Namespace)
+		}
+		_, err := resourceInterface.Patch(
+			ctx,
+			key.Name,
+			types.ApplyPatchType,
+			manifest,
+			metav1.PatchOptions{
+				Force:        ptr.To(force),
+				FieldManager: fieldManager,
+			})
+		return err
+	}
+}
+
+func defaultManagedFunc(oc *clients.OperatorClient) managedFunc {
+	return func() (bool, error) {
+		operatorSpec, _, _, err := oc.GetOperatorState()
+		if err != nil {
+			return false, err
+		}
+		return management.IsOperatorManaged(operatorSpec.ManagementState) && (operatorSpec.ManagementState != operatorv1.Removed || management.IsOperatorNotRemovable()), nil
+	}
+}
+
+func defaultShouldUpdateFunc() shouldUpdateFunc {
+	return func(manifest []byte, u *unstructured.Unstructured) (bool, error) {
+		required, _, err := scheme.Codecs.UniversalDecoder().Decode(manifest, nil, &unstructured.Unstructured{})
+		if err != nil {
+			return false, fmt.Errorf("decoding manifest: %w", err)
+		}
+
+		// we want the inverse of deep derivative to dictate if we should update.
+		// i.e If it is not deep derivative, meaning the requirements are not met, we should update
+		return !equality.Semantic.DeepDerivative(required.(*unstructured.Unstructured).UnstructuredContent(), u.UnstructuredContent()), nil
+	}
+}
+
+// applyFunc is a function that is used and expected to perform a
+// server side apply operation. if there is an error during the apply operation,
+// it is returned.
+type applyFunc func(context.Context, types.NamespacedName, string, bool, schema.GroupVersionResource, []byte) error
+
+// managedFunc is a function that returns whether or not the operator
+// is managed. Any errors encountered while evaluating if this operator is
+// managed are returned.
+type managedFunc func() (bool, error)
+
+// shouldUpdateFunc is a function that takes in raw bytes, expected to be a valid
+// Kubernetes resource YAML, and compares it to an *unstructured.Unstructured.
+// if there is a difference between the raw bytes and the *unstructured.Unstructured
+// that warrants an update to the *unstructured.Unstructured, this function will return
+// true with a nil error. Any errors encountered during the evaluation will be returned.
+type shouldUpdateFunc func([]byte, *unstructured.Unstructured) (bool, error)
+
+// getCatalogFunc is a function that gets a ClusterCatalog resource using the provided
+// key. It is returned as a runtime.Object. Any errors encountered during the process
+// of fetching the ClusterCatalog resource will be returned.
+type getCatalogFunc func(types.NamespacedName) (runtime.Object, error)
```

Comment:
Since this is already an abstraction, maybe make this return a concrete type?
```suggestion
type getCatalogFunc func(types.NamespacedName) (*unstructured.Unstructured, error)
```

Reply:
In this particular case, I generally prefer to use the lowest primitive type and put any type conversion responsibilities on the caller. This makes this code the most extensible because AFAIK all Go types that are used to represent a Kubernetes resource implement the `runtime.Object` interface meaning that you could swap in and out any client that returns a Go type for a Kubernetes resource.

I do think that I should go back and make the signature of the `shouldUpdateFunc` type to take in a `runtime.Object` instead of an `*unstructured.Unstructured` for a similar reason.

Reply:
This should be addressed by the more generic approach taken in https://github.com/openshift/cluster-olm-operator/pull/73/commits/5b237d4e4d462f19ff7571661b839393bde6b676